examples/text_labels_and_annotations/custom_legends.py

```diff
@@ -0,0 +1,56 @@
+"""
+========================
+Composing Custom Legends
+========================
+
+Composing custom legends piece-by-piece.
+
+Sometimes you don't want a legend that is explicitly tied to data that
+you have plotted. For example, say you have plotted 10 lines, but don't
+want a legend item to show up for each one. If you simply plot the lines
+and call ``ax.legend()``, you will get the following:
+"""
+# sphinx_gallery_thumbnail_number 2
+from matplotlib import rcParams, cycler
+import matplotlib.pyplot as plt
+import numpy as np
+
+N = 10
+data = [np.logspace(0, 1, 100) + np.random.randn(100) + ii for ii in range(N)]
+data = np.array(data).T
+cmap = plt.cm.coolwarm
+rcParams['axes.prop_cycle'] = cycler(color=cmap(np.linspace(0, 1, N)))
+
+fig, ax = plt.subplots()
+lines = ax.plot(data)
+ax.legend(lines, loc=(1.05, .05), )
+
+##############################################################################
+# In this case, we can compose a legend using Matplotlib objects that aren't
+# explicitly tied to the data that was plotted. For example:
+
+from matplotlib.lines import Line2D
+custom_lines = [Line2D([0], [0], color='b', lw=4),
+                Line2D([0], [0], color='grey', lw=4),
+                Line2D([0], [0], color='r', lw=4)]
+
+fig, ax = plt.subplots()
+lines = ax.plot(data)
+ax.legend(custom_lines, ['Cold', 'Medium', 'Hot'], loc=(1.05, .8),)
+
+
+###############################################################################
+# There are many other Matplotlib objects that can be used in this way. Below
+# is a list of common ones.
+
+from matplotlib.patches import Patch
+from matplotlib.lines import Line2D
+
+legend_elements = {'Line': Line2D([0], [0], color='b', lw=4),
+                   'Scatter': Line2D([0], [0], marker='o', color='w',
+                                     markerfacecolor='g', markersize=15),
+                   'Color Patch': Patch(facecolor='orange', edgecolor='r')}
+fig, ax = plt.subplots()
+ax.legend(legend_elements.values(), legend_elements.keys(), loc=(1.05, .8),)
```

Comment:
Note that the order of iteration over values()/keys() is not guaranteed (in fact it *will* change between iterations on Py3.4/3.5), so please use lists.

Also I think it is *much* better style to set the label on the artist itself:
```
legend_elements = [
    Line2D(..., label="foo"),
    Patch(..., label="bar)
]
ax.legend(legend_elements)
```

Reply:
Good point re: lists

for setting the label on the artist, IIRC doing that didn't automatically add labels when I called `ax.legend` but I could be wrong...

Reply:
ya, when I do

```
legend_elements = [Line2D([0], [0], color='b', lw=4, label='Line'),
                   Line2D([0], [0], marker='o', color='w', label='Scatter',
                                     markerfacecolor='g', markersize=15),
                   Patch(facecolor='orange', edgecolor='r',
                         label='Color Patch')]
fig, ax = plt.subplots()
ax.legend(legend_elements)
```
the result is
![image](https://user-images.githubusercontent.com/1839645/31643083-60b99606-b2a3-11e7-9fb8-793adb139fb7.png)

is that a bug?

Reply:
Of course :/, this has to be `plt.legend(handles=[...])`, otherwise it thinks you're passing *labels* for some nonexistent artists.  Stupid API, refactoring it is somewhere on my (thick) todo.

Reply:
Well we could do it here: https://github.com/matplotlib/matplotlib/pull/9324  I agree its a backwards API.

Reply:
I would suggest having this on a separate PR.  Basically, if only one positional argument is passed and it's a list of artists, then consider it as handles instead of labels.

Reply:
in the meantime, I'll just use `plt`

Reply:
hmmm, I actually just tried it w/ `plt` and had the same behavior...

Reply:
ah nvm, I see what you mean about `handles=`. That works for `ax.legend` as well actually

Reply:
Right - the positional argument is for labels believe it or not.  i.e. `ax.legend({'a', 'b', 'c'}) will label the first three artists it finds on `ax` 'a', 'b', and 'c'.  I guess artists are stored in the order they are drawn, but it seems like a strange choice to me.  Once #9324 gets merged it'll be easy to check for a list of handles instead and make the call you were trying to make work.